File: c/graphs/zero_one_bfs.c
#include <stdbool.h>

typedef struct {
    int row;
    int col;
} Cell;

int zero_one_bfs(
    int width,
    int height,
    bool walls[height][width],
    int weights[height][width],
    Cell start,
    Cell target,
    Cell path[]
) {
    const int total = width * height;
    const int infinity = total + 1;
    int distance[height][width];
    Cell parent[height][width];
    Cell deque[total * 2];
    int front = total;
    int back = total + 1;

    for (int row = 0; row < height; row++) {
        for (int col = 0; col < width; col++) {
            distance[row][col] = infinity;
            parent[row][col] = (Cell){-1, -1};
        }
    }

    distance[start.row][start.col] = 0;
    deque[back++] = start;

    while (front < back) {
        Cell current = deque[front++];

        if (current.row == target.row && current.col == target.col) {
            int length = 0;
            for (Cell step = target; step.row != -1; step = parent[step.row][step.col]) {
                path[length++] = step;
                if (step.row == start.row && step.col == start.col) {
                    break;
                }
            }

            for (int left = 0; left < length / 2; left++) {
                Cell temp = path[left];
                path[left] = path[length - left - 1];
                path[length - left - 1] = temp;
            }
            return length;
        }

        Cell neighbors[4] = {
            {current.row - 1, current.col},
            {current.row, current.col + 1},
            {current.row + 1, current.col},
            {current.row, current.col - 1},
        };

        for (int index = 0; index < 4; index++) {
            Cell next = neighbors[index];
            if (next.row < 0 || next.row >= height || next.col < 0 || next.col >= width) {
                continue;
            }
            if (walls[next.row][next.col]) {
                continue;
            }

            int weight = weights[next.row][next.col];
            int next_distance = distance[current.row][current.col] + weight;
            if (next_distance >= distance[next.row][next.col]) {
                continue;
            }

            distance[next.row][next.col] = next_distance;
            parent[next.row][next.col] = current;
            if (weight == 0) {
                deque[--front] = next;
            } else {
                deque[back++] = next;
            }
        }
    }

    return 0;
}

File: src/graphs/zeroOneBfs.ts
import { zeroOneBfs, createZeroOneBfsInitialStep } from "./algorithms/zeroOneBfs.js";
import { GraphController } from "./visualizer/graphController.js";
import { GraphData, GridPoint } from "./visualizer/types.js";
import { GraphVisualizer, GridEditMode } from "./visualizer/graphVisualizer.js";
import { createClusterGrid } from "./visualizer/gridGenerator.js";
import { updateGrid } from "./visualizer/gridEditor.js";
import { initCodeLoader } from "../ui/codeLoader.js";
import { initLearnMore } from "../ui/navigation.js";

function key(point: GridPoint): string {
    return `${point.row},${point.col}`;
}

function withWeights(graph: GraphData, seed: number): GraphData {
    const walls = new Set(graph.walls.map(key));
    const weights: Record<string, 0 | 1> = {};

    for (let row = 0; row < graph.height; row++) {
        for (let col = 0; col < graph.width; col++) {
            const point = { row, col };
            if (!walls.has(key(point))) {
                weights[key(point)] = (row * 7 + col * 11 + seed) % 4 === 0 ? 1 : 0;
            }
        }
    }

    weights[key(graph.start)] = 0;
    return { ...graph, weights };
}

const baseGrids: GraphData[] = [
    {
        width: 10,
        height: 7,
        start: { row: 4, col: 0 },
        target: { row: 0, col: 9 },
        walls: [
            { row: 0, col: 4 }, { row: 1, col: 1 }, { row: 1, col: 4 }, { row: 1, col: 7 },
            { row: 2, col: 1 }, { row: 2, col: 7 }, { row: 3, col: 3 }, { row: 3, col: 4 },
            { row: 4, col: 4 }, { row: 4, col: 6 }, { row: 5, col: 1 }, { row: 5, col: 2 },
            { row: 5, col: 6 }, { row: 6, col: 4 }
        ]
    },
    {
        width: 10,
        height: 7,
        start: { row: 6, col: 0 },
        target: { row: 0, col: 9 },
        walls: [
            { row: 0, col: 2 }, { row: 0, col: 3 }, { row: 1, col: 5 }, { row: 1, col: 6 },
            { row: 2, col: 1 }, { row: 2, col: 6 }, { row: 3, col: 1 }, { row: 3, col: 3 },
            { row: 3, col: 4 }, { row: 3, col: 8 }, { row: 4, col: 4 }, { row: 4, col: 8 },
            { row: 5, col: 2 }, { row: 5, col: 3 }, { row: 5, col: 6 }
        ]
    }
].map((graph, index) => withWeights(graph, index + 1));

function createRandomWeightedGrid(): GraphData {
    const start = { row: 6, col: 0 };
    const target = { row: 0, col: 9 };
    const graph = createClusterGrid(10, 7, start, target);
    return withWeights(graph, Math.floor(Math.random() * 40));
}

document.addEventListener("DOMContentLoaded", () => {
    initLearnMore();
    initCodeLoader("graphs", "zero_one_bfs", {
        rootSelector: "#standardCodeSection",
        defaultLanguage: "python",
        moreInfoSelector: "#standardMoreInfo"
    });
    initCodeLoader("graphs", "zero_one_bfs_production", {
        rootSelector: "#productionCodeSection",
        defaultLanguage: "python"
    });

    const visualizer = new GraphVisualizer("zeroOneBfsContainer");
    const mapSelect = document.getElementById("mapSelect") as HTMLSelectElement;
    const randomBtn = document.getElementById("randomBtn") as HTMLButtonElement;
    const resetBtn = document.getElementById("resetBtn") as HTMLButtonElement;
    const editModeSelect = document.getElementById("editModeSelect") as HTMLSelectElement;
    const playBtn = document.getElementById("playBtn") as HTMLButtonElement;
    const pauseBtn = document.getElementById("pauseBtn") as HTMLButtonElement;
    const stepBtn = document.getElementById("stepBtn") as HTMLButtonElement;
    const speedRange = document.getElementById("speedRange") as HTMLInputElement;

    let graph = baseGrids[0];
    const controller = new GraphController(zeroOneBfs(graph), visualizer);
    controller.reset(zeroOneBfs(graph), createZeroOneBfsInitialStep(graph));
    controller.setSpeed(Number(speedRange.value));

    visualizer.setGridEditHandler((point, mode) => {
        controller.pause();
        graph = updateGrid(graph, point, mode);
        controller.reset(zeroOneBfs(graph), createZeroOneBfsInitialStep(graph));
        mapSelect.value = "custom";
    });

    function reset() {
        const selectedIndex = Number(mapSelect.value);
        if (Number.isFinite(selectedIndex) && baseGrids[selectedIndex]) graph = baseGrids[selectedIndex];
        controller.reset(zeroOneBfs(graph), createZeroOneBfsInitialStep(graph));
        controller.setSpeed(Number(speedRange.value));
    }

    mapSelect.addEventListener("change", reset);
    randomBtn.addEventListener("click", () => {
        graph = createRandomWeightedGrid();
        mapSelect.value = "custom";
        controller.reset(zeroOneBfs(graph), createZeroOneBfsInitialStep(graph));
    });
    resetBtn.addEventListener("click", reset);
    editModeSelect.addEventListener("change", () => {
        visualizer.setEditMode(editModeSelect.value as GridEditMode);
    });
    playBtn.addEventListener("click", () => controller.play());
    pauseBtn.addEventListener("click", () => controller.pause());
    stepBtn.addEventListener("click", () => controller.step());
    speedRange.addEventListener("input", () => controller.setSpeed(Number(speedRange.value)));
});
